# Projeto 3 - Ciência dos Dados

#  Projeto UFC

Nome: Victor Vergara Arcoverde de Albuquerque Cavalcanti

Nome: Edgard Neto

Nome: Gabriel Yamashita

Nome: Henrique Mualem Marti



  ___
## Objetivo:

### O objetivo desse projeto é fazer um machine learning para poder prever qual lutador ganhará uma luta do UFC baseado no seu histórico. Para isso serão usados os dados de todas as lutas do UFC (mais de 5 mil) a fim de descobrir quais os fatores dos lutadores que impactam mais no resultado das lutas. 

[Database utilizado](https://www.kaggle.com/rajeevw/ufcdata#data.csv)

 ___
## Preparando o ambiente no jupyter:

### Imports:

In [1]:
import math
import os.path
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
import json
import random
import statsmodels.api as sm
import seaborn as sns

from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics import accuracy_score
from sklearn.naive_bayes import MultinomialNB

from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

### Trabalhando com os Excels:

In [2]:
data = pd.read_excel("data.xlsx")
preproc = pd.read_excel("preprocessed_data.xlsx")
fighter = pd.read_excel("raw_fighter_details.xlsx")
total = pd.read_excel("raw_total_fight_data.xlsx")
data2 = data
data.head(2)

R_fighter      B_fighter          Referee  date  \
0          Henry Cejudo  Marlon Moraes     Marc Goddard  2019   
1  Valentina Shevchenko    Jessica Eye  Robert Madrigal  2019   

                 location Winner  title_bout       weight_class  no_of_rounds  \
0  Chicago, Illinois, USA    Red        True       Bantamweight             5   
1  Chicago, Illinois, USA    Red        True  Women's Flyweight             5   

   B_current_lose_streak  ...  R_win_by_KO/TKO  R_win_by_Submission  \
0                      0  ...                2                    0   
1                      0  ...                0                    2   

   R_win_by_TKO_Doctor_Stoppage  R_wins  R_Stance  R_Height_cms  R_Reach_cms  \
0                             0       8  Orthodox        162.56       162.56   
1                             0       5  Southpaw        165.10       167.64   

   R_Weight_lbs  B_age  R_age  
0         135.0   31.0   32.0  
1         125.0   32.0   31.0  

[2 rows x 145 columns]

In [3]:
data.weight_class = data.weight_class.astype('category')
data.Winner = data.Winner.astype('category')

In [4]:
data.Winner.value_counts()

Red     3470
Blue    1591
Draw      83
Name: Winner, dtype: int64

In [5]:
data_heavy = data.loc[(data.weight_class=='Heavyweight'),:]
bool_to_number = {False: 0, True: 1}
string_to_number = {'Blue': 0, 'Red': 1, 'Draw': 2}
data_heavy['title_bout'] = data_heavy['title_bout'].map(bool_to_number)
data_heavy['Winner'] = data_heavy['Winner'].map(string_to_number)

C:\Users\Maulem\Spyder\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.
C:\Users\Maulem\Spyder\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """


#### Blue = 0
#### Red = 1

In [6]:
data_heavy.head(2)

R_fighter       B_fighter          Referee  date  \
4   Tai Tuivasa   Blagoy Ivanov  Dan Miragliotta  2019   
54  Walt Harris  Serghei Spivac      Jerin Valel  2019   

                   location  Winner  title_bout weight_class  no_of_rounds  \
4    Chicago, Illinois, USA       0           0  Heavyweight             3   
54  Ottawa, Ontario, Canada       1           0  Heavyweight             3   

    B_current_lose_streak  ...  R_win_by_KO/TKO  R_win_by_Submission  \
4                       0  ...                2                    0   
54                      0  ...                4                    0   

    R_win_by_TKO_Doctor_Stoppage  R_wins  R_Stance  R_Height_cms  R_Reach_cms  \
4                              0       3  Southpaw        187.96       190.50   
54                             0       4  Southpaw        195.58       195.58   

    R_Weight_lbs  B_age  R_age  
4          264.0   32.0   26.0  
54         250.0   24.0   35.0  

[2 rows x 145 columns]

In [7]:
data_heavy_util = data_heavy.drop(['Referee','date','location'], axis=1)
#dados que não se relacionam com os lutadores ou seus resultado

In [8]:
data_heavy_util.head()

R_fighter          B_fighter  Winner  title_bout weight_class  \
4       Tai Tuivasa      Blagoy Ivanov       0           0  Heavyweight   
54      Walt Harris     Serghei Spivac       1           0  Heavyweight   
60    Arjan Bhullar         Juan Adams       1           0  Heavyweight   
63       Greg Hardy  Dmitrii Smoliakov       1           0  Heavyweight   
69  Andrei Arlovski      Augusto Sakai       0           0  Heavyweight   

    no_of_rounds  B_current_lose_streak  B_current_win_streak  B_draw  \
4              3                      0                     1       0   
54             3                      0                     0       0   
60             3                      0                     1       0   
63             3                      2                     0       0   
69             3                      0                     1       0   

    B_avg_BODY_att  ...  R_win_by_KO/TKO  R_win_by_Submission  \
4             17.0  ...                2                    0   
54             NaN  ...                4                    0   
60             8.0  ...                0                    0   
63             5.0  ...                0                    0   
69            35.0  ...                9                    2   

    R_win_by_TKO_Doctor_Stoppage  R_wins  R_Stance  R_Height_cms  R_Reach_cms  \
4                              0       3  Southpaw        187.96       190.50   
54                             0       4  Southpaw        195.58       195.58   
60                             0       2  Orthodox        185.42       190.50   
63                             0       0  Orthodox        195.58       203.20   
69                             0      16  Orthodox        190.50       195.58   

    R_Weight_lbs  B_age  R_age  
4          264.0   32.0   26.0  
54         250.0   24.0   35.0  
60         245.0   27.0   32.0  
63         265.0   36.0   30.0  
69         240.0   27.0   40.0  

[5 rows x 142 columns]

In [9]:
data_heavy_util.dropna(inplace=True)
data_heavy_util.head()

R_fighter          B_fighter  Winner  title_bout weight_class  \
4        Tai Tuivasa      Blagoy Ivanov       0           0  Heavyweight   
60     Arjan Bhullar         Juan Adams       1           0  Heavyweight   
63        Greg Hardy  Dmitrii Smoliakov       1           0  Heavyweight   
69   Andrei Arlovski      Augusto Sakai       0           0  Heavyweight   
75  Alistair Overeem    Aleksei Oleinik       1           0  Heavyweight   

    no_of_rounds  B_current_lose_streak  B_current_win_streak  B_draw  \
4              3                      0                     1       0   
60             3                      0                     1       0   
63             3                      2                     0       0   
69             3                      0                     1       0   
75             5                      0                     2       0   

    B_avg_BODY_att  ...  R_win_by_KO/TKO  R_win_by_Submission  \
4            17.00  ...                2                    0   
60            8.00  ...                0                    0   
63            5.00  ...                0                    0   
69           35.00  ...                9                    2   
75            4.25  ...                6                    0   

    R_win_by_TKO_Doctor_Stoppage  R_wins  R_Stance  R_Height_cms  R_Reach_cms  \
4                              0       3  Southpaw        187.96       190.50   
60                             0       2  Orthodox        185.42       190.50   
63                             0       0  Orthodox        195.58       203.20   
69                             0      16  Orthodox        190.50       195.58   
75                             0       9  Orthodox        193.04       203.20   

    R_Weight_lbs  B_age  R_age  
4          264.0   32.0   26.0  
60         245.0   27.0   32.0  
63         265.0   36.0   30.0  
69         240.0   27.0   40.0  
75         265.0   41.0   38.0  

[5 rows x 142 columns]

#### Tirando os dados categóricos:

In [10]:
categoricas = [
    'R_fighter', 
    'B_fighter', 
    'weight_class', 
    'R_Stance', 
    'B_Stance', 
]

data_heavy_cat = data_heavy_util[categoricas].astype('category')
data_heavy_num = data_heavy_util.drop(categoricas, axis=1).astype('float')

In [11]:
X = data_heavy_num.drop('Winner', axis=1)
y = data_heavy_num['Winner']

### Separando os dados em testes e treinamento

In [12]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25)

In [13]:

model = RandomForestClassifier(n_estimators=10000)

model.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10000, n_jobs=None,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

In [14]:
#Fatores que tem o maior peso na decisão da vitória
j=1
lista_j=list()
for i, f in sorted(list(zip(model.feature_importances_, X_train.columns)), reverse=True):
    print( j,'°:',f'{f}: {i}')
    a=str(j)+'°'
    lista_j.append(a)
    j+=1


1 °: R_avg_opp_TOTAL_STR_landed: 0.015668803235770738
2 °: R_avg_opp_CLINCH_landed: 0.01504517747141978
3 °: R_age: 0.01504369485885921
4 °: R_total_rounds_fought: 0.015023952358479446
5 °: B_avg_BODY_landed: 0.01380260122499282
6 °: B_avg_BODY_att: 0.013224129827308835
7 °: R_avg_opp_HEAD_landed: 0.01320113046082939
8 °: R_avg_opp_CLINCH_att: 0.01285749516478218
9 °: B_avg_HEAD_att: 0.012595021020686865
10 °: R_avg_opp_GROUND_landed: 0.012104874968346512
11 °: B_avg_TD_att: 0.011437434469863025
12 °: R_avg_CLINCH_att: 0.01137497761806011
13 °: B_avg_TOTAL_STR_att: 0.010766299173339088
14 °: B_avg_SIG_STR_att: 0.010601896847073336
15 °: R_avg_CLINCH_landed: 0.01060028374504822
16 °: R_losses: 0.010581294195177117
17 °: B_avg_opp_SIG_STR_pct: 0.010563984533764887
18 °: R_wins: 0.010540486276453884
19 °: R_avg_opp_SIG_STR_pct: 0.010487367902637873
20 °: R_avg_opp_GROUND_att: 0.010399862628803784
21 °: R_avg_opp_BODY_landed: 0.010264401486569062
22 °: B_avg_CLINCH_landed: 0.01021237619837

#### Criando um dataframe com os dados acima:

In [15]:
data={'Fator':X_train.columns ,'Correlação':model.feature_importances_,}
Fator_por_corr=pd.DataFrame(data)
Fator_por_corr=Fator_por_corr.sort_values(by='Correlação', ascending=False)
Fator_por_corr['Grau de Importância']=lista_j
Fator_por_corr = Fator_por_corr.set_index('Grau de Importância')
Fator_por_corr.head()

Fator  Correlação
Grau de Importância                                        
1°                   R_avg_opp_TOTAL_STR_landed    0.015669
2°                      R_avg_opp_CLINCH_landed    0.015045
3°                                        R_age    0.015044
4°                        R_total_rounds_fought    0.015024
5°                            B_avg_BODY_landed    0.013803

#### Com base na correlação, vamos definir a relevância:

In [16]:
def relevancia(df,coluna_nome,coluna_correlacao,acuracia):
    inuteis = []
    uteis = []
    for index,row in df.iterrows():
        if row[coluna_correlacao] >= -acuracia and row[coluna_correlacao] <= acuracia:
            inuteis.append(row[coluna_nome])
        else:
            uteis.append(row[coluna_nome])
    return uteis

In [17]:
uteis = relevancia(Fator_por_corr,'Fator','Correlação',0.01)

In [18]:
data_heavy_util_relevante = data_heavy_util.loc[:,uteis]
data_heavy_util_relevante.head()

R_avg_opp_TOTAL_STR_landed  R_avg_opp_CLINCH_landed  R_age  \
4                    27.750000                 3.500000   26.0   
60                   42.333333                 5.333333   32.0   
63                   16.000000                 1.000000   30.0   
69                   34.821429                 3.035714   40.0   
75                   28.533333                 4.200000   38.0   

    R_total_rounds_fought  B_avg_BODY_landed  B_avg_BODY_att  \
4                       7             14.500           17.00   
60                      8              8.000            8.00   
63                      2              3.500            5.00   
69                     59             21.000           35.00   
75                     29              3.875            4.25   

    R_avg_opp_HEAD_landed  R_avg_opp_CLINCH_att  B_avg_HEAD_att  \
4               14.000000              4.500000           184.5   
60              12.333333              8.000000           131.0   
63              12.000000              2.000000            15.5   
69              15.142857              4.964286           169.0   
75              13.333333              6.466667            28.5   

    R_avg_opp_GROUND_landed  ...  R_losses  B_avg_opp_SIG_STR_pct  R_wins  \
4                  2.750000  ...         1                 0.4300       3   
60                 0.666667  ...         1                 0.5300       2   
63                 3.000000  ...         1                 0.4300       0   
69                 5.357143  ...        12                 0.4400      16   
75                 3.466667  ...         6                 0.4775       9   

    R_avg_opp_SIG_STR_pct  R_avg_opp_GROUND_att  R_avg_opp_BODY_landed  \
4                0.397500              7.750000               4.750000   
60               0.416667              1.333333               7.666667   
63               0.400000              3.000000               0.000000   
69               0.389286              7.964286               3.750000   
75               0.437333              4.133333               3.533333   

    B_avg_CLINCH_landed  R_avg_SIG_STR_pct  B_avg_CLINCH_att  R_avg_BODY_att  
4                 2.000           0.545000             2.500        7.750000  
60               13.000           0.510000            15.000        2.000000  
63                3.500           0.450000             5.500        5.000000  
69               30.000           0.453929            38.000        6.607143  
75                4.125           0.688667             5.375       13.466667  

[5 rows x 25 columns]

___
## Verificando a performance


In [19]:
X_train.columns

Index(['title_bout', 'no_of_rounds', 'B_current_lose_streak',
       'B_current_win_streak', 'B_draw', 'B_avg_BODY_att', 'B_avg_BODY_landed',
       'B_avg_CLINCH_att', 'B_avg_CLINCH_landed', 'B_avg_DISTANCE_att',
       ...
       'R_win_by_Decision_Unanimous', 'R_win_by_KO/TKO', 'R_win_by_Submission',
       'R_win_by_TKO_Doctor_Stoppage', 'R_wins', 'R_Height_cms', 'R_Reach_cms',
       'R_Weight_lbs', 'B_age', 'R_age'],
      dtype='object', length=136)

In [20]:
model.feature_importances_

array([0.00037982, 0.0008378 , 0.00198755, 0.00448719, 0.        ,
       0.01322413, 0.0138026 , 0.01008011, 0.01021238, 0.0087328 ,
       0.00796997, 0.00882968, 0.00836895, 0.01259502, 0.00959801,
       0.00548468, 0.00942306, 0.00854325, 0.00697032, 0.00478324,
       0.0106019 , 0.00954819, 0.0079553 , 0.00416357, 0.01143743,
       0.00673843, 0.00649148, 0.0107663 , 0.00992713, 0.00509203,
       0.00503317, 0.00712215, 0.0068983 , 0.00851778, 0.00804232,
       0.00736991, 0.00763647, 0.00787396, 0.00826296, 0.00833778,
       0.00742492, 0.00454079, 0.00816121, 0.00735323, 0.00608274,
       0.00324957, 0.00811884, 0.00726743, 0.01056398, 0.00470418,
       0.00962376, 0.00620242, 0.00765156, 0.00767997, 0.00830097,
       0.00638697, 0.00931229, 0.00166128, 0.00040887, 0.00176182,
       0.00167487, 0.00461331, 0.00141758, 0.00060273, 0.00418817,
       0.00547292, 0.00754578, 0.00745844, 0.00366894, 0.00394249,
       0.        , 0.01007129, 0.00876238, 0.01137498, 0.01060

In [21]:
y_pred = model.predict(X_test)

In [22]:
print(accuracy_score(y_test, y_pred))

0.5757575757575758


In [23]:
y_pred

array([1., 1., 0., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 1., 0., 1., 1., 1., 1., 1.,
       1., 0., 0., 1., 1., 1., 1., 0., 1., 0., 1., 1., 1., 0., 1.])

In [24]:
y_test.value_counts(True)

1.0    0.590909
0.0    0.393939
2.0    0.015152
Name: Winner, dtype: float64

In [25]:
data_heavy_num.shape

(264, 137)

### Conclusão desse teste:

Tendo um acurácia de 0.6667 não é um bom resultado, visto que o modelo praticamente sempre tem como resultado o vermelho como vencedor, e como a probabilidade do lutador vermelho ganhar é de 0.6212112 ele praticamente só acerta os que o vencedor é o vermelho e erra os que o azul é o vencedor.

Assim é possível concluir que é necessário desconsiderar algumas variáveis para tentar melhorar a acurácia.

___
## Montando a Regressão Logística


## !Ainda é necessário tirar os fatores que foram julgados ter um baixo grau de importância pelo teste anterior!

In [26]:
def preparo(X,Y):
    X_cp = sm.add_constant(X)
    model = sm.OLS(Y,X_cp,missing='drop')
    results = model.fit()
    return results

In [27]:
Y = data_heavy_num["Winner"]
data_heavy_sem_Winner=data_heavy_num.drop('Winner',axis=1) 
X=data_heavy_sem_Winner
#np.asarray(X)

In [28]:
result = preparo(X,Y)
result.summary()

C:\Users\Maulem\Spyder\lib\site-packages\numpy\core\fromnumeric.py:2389: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Winner   R-squared:                       0.515
Model:                            OLS   Adj. R-squared:                  0.120
Method:                 Least Squares   F-statistic:                     1.305
Date:                Mon, 18 Nov 2019   Prob (F-statistic):             0.0635
Time:                        22:19:35   Log-Likelihood:                -100.66
No. Observations:                 264   AIC:                             439.3
Df Residuals:                     145   BIC:                             864.9
Df Model:                         118                                         
Covariance Type:            nonrobust                                         
================================================================================================
                                   coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------
const                            4.6535      2.927      1.590      0.114      -1.131      10.438
title_bout                      -0.0643      0.192     -0.334      0.739      -0.445       0.316
no_of_rounds                     0.0958      0.062      1.533      0.127      -0.028       0.219
B_current_lose_streak           -0.2204      0.073     -3.002      0.003      -0.366      -0.075
B_current_win_streak            -0.0361      0.037     -0.983      0.327      -0.109       0.037
B_draw                        1.562e-13   4.62e-14      3.377      0.001    6.48e-14    2.48e-13
B_avg_BODY_att                  -0.0701      0.040     -1.752      0.082      -0.149       0.009
B_avg_BODY_landed                0.0788      0.049      1.623      0.107      -0.017       0.175
B_avg_CLINCH_att                 0.0259      0.025      1.025      0.307      -0.024       0.076
B_avg_CLINCH_landed             -0.0234      0.035     -0.669      0.504      -0.092       0.046
B_avg_DISTANCE_att              -0.0151      0.014     -1.059      0.291      -0.043       0.013
B_avg_DISTANCE_landed            0.0141      0.019      0.727      0.469      -0.024       0.053
B_avg_GROUND_att                -0.0245      0.023     -1.059      0.291      -0.070       0.021
B_avg_GROUND_landed              0.0242      0.032      0.764      0.446      -0.038       0.087
B_avg_HEAD_att                  -0.0283      0.022     -1.283      0.201      -0.072       0.015
B_avg_HEAD_landed                0.0352      0.028      1.266      0.208      -0.020       0.090
B_avg_KD                         0.3686      0.193      1.906      0.059      -0.014       0.751
B_avg_LEG_att                    0.0846      0.059      1.436      0.153      -0.032       0.201
B_avg_LEG_landed                -0.0990      0.071     -1.386      0.168      -0.240       0.042
B_avg_PASS                       0.1028      0.075      1.375      0.171      -0.045       0.251
B_avg_REV                        0.7377      0.295      2.497      0.014       0.154       1.322
B_avg_SIG_STR_att               -0.0137      0.023     -0.606      0.546      -0.059       0.031
B_avg_SIG_STR_landed             0.0150      0.027      0.551      0.583      -0.039       0.069
B_avg_SIG_STR_pct               -1.5196      0.575     -2.644      0.009      -2.656      -0.384
B_avg_SUB_ATT                   -0.1195      0.137     -0.871      0.385      -0.391       0.152
B_avg_TD_att                    -0.0131      0.048     -0.274      0.784      -0.107       0.081
B_avg_TD_landed                 -0.0729      0.104     -0.701      0.484      -0.278       0.133
B_avg_TD_pct                     0.3372      0.275      1.228      0.222      -0.206       0.880
B_avg_TOTAL_STR_att              0.0405      0.026      1.568      0.119    

In [29]:
'''data_heavy_blue_winners = data_heavy.loc[(data_heavy.Winner==0),:]
data_heavy_blue_winners = data_heavy_blue_winners.drop(['Winner'], axis=1)
data_heavy_red_winners = data_heavy.loc[(data_heavy.Winner==1),:]
data_heavy_red_winners = data_heavy_red_winners.drop(['Winner'], axis=1)'''

"data_heavy_blue_winners = data_heavy.loc[(data_heavy.Winner==0),:]\ndata_heavy_blue_winners = data_heavy_blue_winners.drop(['Winner'], axis=1)\ndata_heavy_red_winners = data_heavy.loc[(data_heavy.Winner==1),:]\ndata_heavy_red_winners = data_heavy_red_winners.drop(['Winner'], axis=1)"

In [30]:
'''blue_winners_correlation = data_heavy_blue_winners.corr()
blue_winners_correlation.head()'''

'blue_winners_correlation = data_heavy_blue_winners.corr()\nblue_winners_correlation.head()'

In [31]:
'''red_winners_correlation = data_heavy_red_winners.corr()
red_winners_correlation.head()'''

'red_winners_correlation = data_heavy_red_winners.corr()\nred_winners_correlation.head()'

___
## Concluindo

   ___
## Aperfeiçoamento:


   ___
## Referências

[Como usar a biblioteca Scikit-lear](https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression)

[Como funciona o Random Forest](https://towardsdatascience.com/understanding-random-forest-58381e0602d2)

[Referencia do Random Forest Classifier](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html)

[Como funciona regressão linear](https://www.saedsayad.com/logistic_regression.htm)

[Como usar a Regressão Linear](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html)



In [32]:
#criando um novo dataframe
lista_locB = ["B_fighter", "weight_class","B_Height_cms","B_Reach_cms","B_Weight_lbs","B_age"]
lista_locR = ["R_fighter", "weight_class","R_Height_cms","R_Reach_cms","R_Weight_lbs","R_age"]
data_red = data2.loc[: , lista_locR]
data_blue = data2.loc[: , lista_locB]

data_blue.columns = ["Fighter","Weight_class","Height_cms","Reach_cms","Weight_lbs","Age"]
data_red.columns = ["Fighter","Weight_class","Height_cms","Reach_cms","Weight_lbs","Age"]

data_red

Fighter           Weight_class  Height_cms  Reach_cms  \
0              Henry Cejudo           Bantamweight      162.56     162.56   
1      Valentina Shevchenko      Women's Flyweight      165.10     167.64   
2             Tony Ferguson            Lightweight      180.34     193.04   
3             Jimmie Rivera           Bantamweight      162.56     172.72   
4               Tai Tuivasa            Heavyweight      187.96     190.50   
5            Tatiana Suarez    Women's Strawweight      165.10     167.64   
6         Aljamain Sterling           Bantamweight      170.18     180.34   
7     Karolina Kowalkiewicz    Women's Strawweight      160.02     162.56   
8             Ricardo Lamas          Featherweight      172.72     180.34   
9               Yan Xiaonan    Women's Strawweight      165.10     160.02   
10              Bevon Lewis           Middleweight      190.50     200.66   
11           Eddie Wineland           Bantamweight      170.18     175.26   
12        Katlyn Chookagian      Women's Flyweight      175.26     172.72   
13     Alexander Gustafsson      Light Heavyweight      195.58     200.66   
14              Jimi Manuwa      Light Heavyweight      185.42     200.66   
15         Makwan Amirkhani          Featherweight      177.80     182.88   
16           Damir Hadzovic            Lightweight      175.26     177.80   
17            Daniel Teymur          Featherweight      165.10     175.26   
18             Rostem Akman           Welterweight      177.80     182.88   
19            Tonya Evinger   Women's Bantamweight      170.18     177.80   
20               Stevie Ray            Lightweight      177.80     177.80   
21                Nick Hein            Lightweight      167.64     167.64   
22              Bea Malecki   Women's Bantamweight      175.26     187.96   
23             Darko Stosic      Light Heavyweight      182.88     193.04   
24             Joel Alvarez            Lightweight      190.50     195.58   
25         Rafael Dos Anjos           Welterweight      172.72     177.80   
26    Antonio Carlos Junior           Middleweight      187.96     200.66   
27           Megan Anderson  Women's Featherweight      182.88     182.88   
28            Vicente Luque           Welterweight      180.34     190.50   
29         Charles Oliveira            Lightweight      177.80     187.96   
...                     ...                    ...         ...        ...   
5114             Guy Mezger            Open Weight      185.42        NaN   
5115           Steve Jennum           Catch Weight      177.80        NaN   
5116           Ken Shamrock            Open Weight      185.42     182.88   
5117           Royce Gracie            Open Weight      185.42        NaN   
5118          Harold Howard            Open Weight      187.96        NaN   
5119           Ken Shamrock            Open Weight      185.42     182.88   
5120          Keith Hackney            Open Weight      180.34        NaN   
5121           Royce Gracie           Catch Weight      185.42        NaN   
5122           Royce Gracie            Open Weight      185.42        NaN   
5123          Patrick Smith            Open Weight      187.96        NaN   
5124           Royce Gracie            Open Weight      185.42        NaN   
5125          Remco Pardoel            Open Weight      190.50        NaN   
5126          Johnny Rhodes            Open Weight      182.88        NaN   
5127          Patrick Smith            Open Weight      187.96        NaN   
5128           Royce Gracie            Open Weight      185.42        NaN   
5129          Jason DeLucia            Open Weight      180.34        NaN   
5130          Remco Pardoel            Open Weight      190.50        NaN   
5131           Orlando Wiet            Open Weight      177.80        NaN   
5132          Frank Hamaker            Open Weight         NaN        NaN   
5133          Johnny Rhodes            Open Weight      182.88        NaN   
5134          Pa

In [35]:
newdataR = pd.DataFrame()


namelistR = []

for n in range(0,5143):
    if data_red.loc[n,:]["Fighter"] not in namelistR:
        namelistR.append(data_red.loc[n,:]["Fighter"])
        newdataR = pd.concat([newdataR,data_red.loc[n,:]], axis=1, join='outer')

newdataR = newdataR.transpose()

newdataR 

C:\Users\Maulem\Spyder\lib\site-packages\ipykernel_launcher.py:9: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  if __name__ == '__main__':


Age                Fighter Height_cms Reach_cms           Weight_class  \
0      32           Henry Cejudo     162.56    162.56           Bantamweight   
1      31   Valentina Shevchenko      165.1    167.64      Women's Flyweight   
2      35          Tony Ferguson     180.34    193.04            Lightweight   
3      29          Jimmie Rivera     162.56    172.72           Bantamweight   
4      26            Tai Tuivasa     187.96     190.5            Heavyweight   
5      28         Tatiana Suarez      165.1    167.64    Women's Strawweight   
6      29      Aljamain Sterling     170.18    180.34           Bantamweight   
7      33  Karolina Kowalkiewicz     160.02    162.56    Women's Strawweight   
8      37          Ricardo Lamas     172.72    180.34          Featherweight   
9      29            Yan Xiaonan      165.1    160.02    Women's Strawweight   
10     28            Bevon Lewis      190.5    200.66           Middleweight   
11     34         Eddie Wineland     170.18    175.26           Bantamweight   
12     30      Katlyn Chookagian     175.26    172.72      Women's Flyweight   
13     32   Alexander Gustafsson     195.58    200.66      Light Heavyweight   
14     39            Jimi Manuwa     185.42    200.66      Light Heavyweight   
15     30       Makwan Amirkhani      177.8    182.88          Featherweight   
16     32         Damir Hadzovic     175.26     177.8            Lightweight   
17     31          Daniel Teymur      165.1    175.26          Featherweight   
18     27           Rostem Akman      177.8    182.88           Welterweight   
19     37          Tonya Evinger     170.18     177.8   Women's Bantamweight   
20     29             Stevie Ray      177.8     177.8            Lightweight   
21     35              Nick Hein     167.64    167.64            Lightweight   
22     27            Bea Malecki     175.26    187.96   Women's Bantamweight   
23     27           Darko Stosic     182.88    193.04      Light Heavyweight   
24     25           Joel Alvarez      190.5    195.58            Lightweight   
25     34       Rafael Dos Anjos     172.72     177.8           Welterweight   
26     29  Antonio Carlos Junior     187.96    200.66           Middleweight   
27     29         Megan Anderson     182.88    182.88  Women's Featherweight   
28     27          Vicente Luque     180.34     190.5           Welterweight   
29     29       Charles Oliveira      177.8    187.96            Lightweight   
...   ...                    ...        ...       ...                    ...   
5035   31         Scott Ferrozzo     180.34       NaN            Heavyweight   
5036   27      Yoshiki Takahashi     180.34       NaN            Lightweight   
5038   31               Don Frye     185.42       NaN           Catch Weight   
5049  NaN         Brian Johnston      190.5       NaN            Open Weight   
5058  NaN            Geza Kalman     185.42       NaN            Open Weight   
5059   37             Dan Severn     187.96       NaN           Catch Weight   
5061  NaN              Mark Hall     182.88       NaN            Open Weight   
5062  NaN           Mark Schultz     175.26       NaN            Open Weight   
5063  NaN          Rafael Carino      203.2       NaN            Open Weight   
5064  NaN            Cal Worsham     180.34       NaN            Open Weight   
5065  NaN          Steve Nelmark     182.88       NaN            Open Weight   
5072  NaN          Paul Varelans      203.2       NaN            Open Weight   
5075   28          Oleg Taktarov     182.88       NaN            Open Weight   
5078   34             Marco Ruas     185.42       NaN            Open Weight   
5086  NaN          Remco Pardoel      190.5       NaN            Open Weight   
5089  NaN           Scott Bessac     193.04       NaN            Open Weight   
5095   31          Patrick Smith     187.96       NaN            Open Weight   
5104  NaN            Todd Medina      177.8       NaN            Open Weight   
5105  

In [36]:
newdataB = pd.DataFrame()


namelistB = []

for n in range(0,5143):
    if data_blue.loc[n,:]["Fighter"] not in namelistB:
        namelistB.append(data_blue.loc[n,:]["Fighter"])
        newdataB = pd.concat([newdataB,data_blue.loc[n,:]], axis=1, join='outer')

newdataB = newdataB.transpose()

newdataB

C:\Users\Maulem\Spyder\lib\site-packages\ipykernel_launcher.py:9: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  if __name__ == '__main__':


Age              Fighter Height_cms Reach_cms           Weight_class  \
0      31        Marlon Moraes     167.64    170.18           Bantamweight   
1      32          Jessica Eye     167.64    167.64      Women's Flyweight   
2      36       Donald Cerrone     185.42    185.42            Lightweight   
3      26             Petr Yan     170.18    170.18           Bantamweight   
4      32        Blagoy Ivanov     180.34    185.42            Heavyweight   
5      33        Nina Ansaroff      165.1    162.56    Women's Strawweight   
6      32         Pedro Munhoz     167.64     165.1           Bantamweight   
7      25         Alexa Grasso      165.1    167.64    Women's Strawweight   
8      31        Calvin Kattar     180.34    182.88          Featherweight   
9      34          Angela Hill     160.02    162.56    Women's Strawweight   
10     28       Darren Stewart     182.88    187.96           Middleweight   
11     35       Grigorii Popov     170.18    172.72           Bantamweight   
12     33    Joanne Calderwood     167.64     165.1      Women's Flyweight   
13     30        Anthony Smith     193.04    193.04      Light Heavyweight   
14     27     Aleksandar Rakic     193.04    198.12      Light Heavyweight   
15     26       Chris Fishgold     172.72    172.72          Featherweight   
16     29      Christos Giagos      177.8    180.34            Lightweight   
17     26          Sung Bin Jo     182.88    182.88          Featherweight   
18     27    Sergey Khandozhko     185.42    187.96           Welterweight   
19     37        Lina Lansberg     170.18     165.1   Women's Bantamweight   
20     39      Leonardo Santos     182.88     190.5            Lightweight   
21     30        Frank Camacho      177.8    185.42            Lightweight   
22     23         Duda Santana     170.18    185.42   Women's Bantamweight   
23     29          Devin Clark     182.88     190.5      Light Heavyweight   
24     24     Danilo Belluardo     182.88    187.96            Lightweight   
25     26            Kevin Lee     175.26    195.58           Welterweight   
26     30         Ian Heinisch     180.34    182.88           Middleweight   
27     28      Felicia Spencer     167.64    172.72  Women's Featherweight   
28     31       Derrick Krantz      177.8    182.88           Welterweight   
29     34            Nik Lentz     172.72    172.72            Lightweight   
...   ...                  ...        ...       ...                    ...   
5097  NaN           John Matua     187.96       NaN            Open Weight   
5101  NaN          Todd Medina      177.8       NaN            Open Weight   
5103  NaN       Ernie Verdicia     180.34       NaN            Open Weight   
5105   20        Andy Anderson     167.64       NaN            Open Weight   
5106  NaN         Asbel Cancio      177.8       NaN            Open Weight   
5108  NaN       Marcus Bossett     185.42       NaN            Open Weight   
5111   25         Melton Bowen     182.88       NaN            Open Weight   
5112   24              Joe Son     162.56       NaN            Open Weight   
5113   51        Ron van Clief      177.8       NaN            Open Weight   
5114  NaN          Jason Fairn     185.42       NaN            Open Weight   
5116  NaN   Felix Lee Mitchell        NaN       NaN            Open Weight   
5118  NaN         Roland Payne     175.26       NaN            Open Weight   
5120   34  Emmanuel Yarborough      203.2       NaN            Open Weight   
5121   30        Patrick Smith     187.96       NaN           Catch Weight   
5123  NaN        Johnny Rhodes     182.88       NaN            Open Weight   
5125  NaN         Orlando Wiet      177.8       NaN            Open Weight   
5126  NaN          Fred Ettish     182.88       NaN            Open Weight   
5127  NaN         Scott Morris      177.8       NaN            Open Weight   
5128  NaN      Minoki Ichihara     170.18       NaN            Open Weight   
5129  NaN          Scott Bak

In [37]:
newdata = pd.concat([newdataB, newdataR], axis=0, join='outer', ignore_index=False, keys=None,levels=None, names=None, verify_integrity=False, copy=True)
newdata = newdata.set_index("Fighter")

newdata

Age Height_cms Reach_cms           Weight_class Weight_lbs
Fighter                                                                      
Marlon Moraes       31     167.64    170.18           Bantamweight        135
Jessica Eye         32     167.64    167.64      Women's Flyweight        125
Donald Cerrone      36     185.42    185.42            Lightweight        155
Petr Yan            26     170.18    170.18           Bantamweight        135
Blagoy Ivanov       32     180.34    185.42            Heavyweight        250
Nina Ansaroff       33      165.1    162.56    Women's Strawweight        115
Pedro Munhoz        32     167.64     165.1           Bantamweight        135
Alexa Grasso        25      165.1    167.64    Women's Strawweight        115
Calvin Kattar       31     180.34    182.88          Featherweight        145
Angela Hill         34     160.02    162.56    Women's Strawweight        115
Darren Stewart      28     182.88    187.96           Middleweight        185
Grigorii Popov      35     170.18    172.72           Bantamweight        135
Joanne Calderwood   33     167.64     165.1      Women's Flyweight        125
Anthony Smith       30     193.04    193.04      Light Heavyweight        205
Aleksandar Rakic    27     193.04    198.12      Light Heavyweight        205
Chris Fishgold      26     172.72    172.72          Featherweight        145
Christos Giagos     29      177.8    180.34            Lightweight        155
Sung Bin Jo         26     182.88    182.88          Featherweight        145
Sergey Khandozhko   27     185.42    187.96           Welterweight        170
Lina Lansberg       37     170.18     165.1   Women's Bantamweight        135
Leonardo Santos     39     182.88     190.5            Lightweight        155
Frank Camacho       30      177.8    185.42            Lightweight        170
Duda Santana        23     170.18    185.42   Women's Bantamweight        145
Devin Clark         29     182.88     190.5      Light Heavyweight        205
Danilo Belluardo    24     182.88    187.96            Lightweight        155
Kevin Lee           26     175.26    195.58           Welterweight        170
Ian Heinisch        30     180.34    182.88           Middleweight        185
Felicia Spencer     28     167.64    172.72  Women's Featherweight        145
Derrick Krantz      31      177.8    182.88           Welterweight        170
Nik Lentz           34     172.72    172.72            Lightweight        155
...                ...        ...       ...                    ...        ...
Scott Ferrozzo      31     180.34       NaN            Heavyweight        323
Yoshiki Takahashi   27     180.34       NaN            Lightweight        199
Don Frye            31     185.42       NaN           Catch Weight        219
Brian Johnston     NaN      190.5       NaN            Open Weight        230
Geza Kalman        NaN     185.42       NaN            Open Weight        200
Dan Severn          37     187.96       NaN           Catch Weight        250
Mark Hall          NaN     182.88       NaN            Open Weight        190
Mark Schultz       NaN     175.26       NaN            Open Weight        200
Rafael Carino      NaN      203.2       NaN            Open Weight        245
Cal Worsham        NaN     180.34       NaN            Open Weight        230
Steve Nelmark      NaN     182.88       NaN            Open Weight        240
Paul Varelans      NaN      203.2       NaN            Open Weight        300
Oleg Taktarov       28     182.88       NaN            Open Weight        210
Marco Ruas          34     185.42       NaN            Open Weight        210
Remco Pardoel      NaN      190.5       NaN            Open Weight        260
Scott Bessac       NaN     193.04       NaN            Open Weight        245
Patrick Smith       31     187.96       NaN            Open Weight        225
Todd Medina        NaN      177.8       NaN            Open Weight        193
Jon Hess           NaN     200.66       N